File: serializers/gson/src/main/java/at/stefangeyer/challonge/serializer/gson/adapter/AttachmentAdapter.java
package at.stefangeyer.challonge.serializer.gson.adapter;

import at.stefangeyer.challonge.model.Attachment;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.time.OffsetDateTime;

/**
 * Type adapter for the [Attachment] class.
 * The received json object is being unpacked.
 *
 * @author Stefan Geyer
 * @version 2018-07-06
 */
public class AttachmentAdapter implements JsonDeserializer<Attachment> {

    @Override
    public Attachment deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonObject e = json.getAsJsonObject().has("match_attachment") ?
                json.getAsJsonObject().get("match_attachment").getAsJsonObject() : json.getAsJsonObject();

        long id = e.get("id").getAsLong();
        long matchId = e.get("match_id").getAsLong();
        long userId = e.get("user_id").getAsLong();
        String description = getOrNull(e, "description") != null ? getOrNull(e, "description").getAsString() : null;
        String url = getOrNull(e, "url") != null ? getOrNull(e, "url").getAsString() : null;
        String originalFileName = getOrNull(e, "original_file_name") != null ? getOrNull(e, "original_file_name").getAsString() : null;
        OffsetDateTime createdAt = context.deserialize(e.get("created_at"), OffsetDateTime.class);
        OffsetDateTime updatedAt = context.deserialize(e.get("updated_at"), OffsetDateTime.class);
        String assetFileName = getOrNull(e, "asset_file_name") != null ? getOrNull(e, "asset_file_name").getAsString() : null;
        String assetContentType = getOrNull(e, "asset_content_type") != null ? getOrNull(e, "asset_content_type").getAsString() : null;
        Long assetFileSize = getOrNull(e, "asset_file_size") != null ? getOrNull(e, "asset_file_size").getAsLong() : null;
        String assetUrl = getOrNull(e, "asset_url") != null ? getOrNull(e, "asset_url").getAsString() : null;
        if (assetUrl != null) {
            assetUrl = "https:" + assetUrl;
        }

        return Attachment.builder().id(id).matchId(matchId).userId(userId).description(description).url(url)
                .originalFileName(originalFileName).createdAt(createdAt).updatedAt(updatedAt).assetFileName(assetFileName)
                .assetContentType(assetContentType).assetFileSize(assetFileSize).assetUrl(assetUrl).build();
    }

    private JsonElement getOrNull(JsonObject o, String key) {
        return !o.has(key) || o.get(key).isJsonNull() ? null : o.get(key);
    }
}

File: serializers/gson/src/main/kotlin/at/stefangeyer/challonge/serializer/gson/adapter/AttachmentAdapter.kt
/*
 * Copyright 2017 Stefan Geyer <stefangeyer at outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.stefangeyer.challonge.serializer.gson.adapter

import at.stefangeyer.challonge.model.Attachment
import at.stefangeyer.challonge.serializer.gson.util.getOrNull
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import java.time.OffsetDateTime

/**
 * Type adapter for the [Attachment] class.
 * The received json object is being unpacked.
 *
 * @author Stefan Geyer
 * @version 2018-07-06
 */
class AttachmentAdapter : JsonDeserializer<Attachment> {

    @Throws(JsonParseException::class)
    override fun deserialize(jsonElement: JsonElement, type: Type, context: JsonDeserializationContext): Attachment {
        val e = jsonElement.asJsonObject

        val id = e.get("id").asLong
        val matchId = e.get("match_id").asLong
        val userId = e.get("user_id").asLong
        val description = e.get("description").getOrNull()?.asString
        val url = e.get("url").getOrNull()?.asString
        val originalFileName = e.get("original_file_name").getOrNull()?.asString
        val createdAt = context.deserialize<OffsetDateTime>(e.get("created_at"), OffsetDateTime::class.java)
        val updatedAt = context.deserialize<OffsetDateTime>(e.get("updated_at"), OffsetDateTime::class.java)
        val assetFileName = e.get("asset_file_name").getOrNull()?.asString
        val assetContentType = e.get("asset_content_type").getOrNull()?.asString
        val assetFileSize = e.get("asset_file_size").getOrNull()?.asLong
        var assetUrl = e.get("asset_url").getOrNull()?.asString
        if (assetUrl != null) assetUrl = "https:$assetUrl"

        return Attachment(
                id = id, matchId = matchId, userId = userId, description = description, url = url, originalFileName = originalFileName,
                createdAt = createdAt, updatedAt = updatedAt, assetFileName = assetFileName, assetContentType = assetContentType,
                assetFileSize = assetFileSize, assetUrl = assetUrl)
    }
}